File: frontend/src/services/api/robot.ts
import { apiClient } from './client';

// 机器人管理相关类型
export interface RobotStatus {
  hardware_available: boolean;
  hardware_status: {
    simulation_available: boolean;
    roslibpy_available: boolean;
    gazebo_running: boolean;
    pybullet_available: boolean;
    pybullet_connected: boolean;
    pybullet_physics_engine: string;
    active_simulation: string;
    robot_model_loaded: boolean;
    physics_engine: string;
    simulation_time: number;
    real_time_factor: number;
  };
  ros_connection: {
    connected: boolean;
    uri: string;
    port: number;
    topics: string[];
    services: string[];
    last_connection: string | null;
    connection_duration: number;
  };
  joint_count: number;
  sensor_count: number;
  control_modes: string[];
  simulation_enabled: boolean;
  real_robot_connected: boolean;
  battery_level: number;
  cpu_temperature: number;
  last_update: string;
  capabilities: Record<string, boolean>;
}

// 多机器人管理相关类型
export enum RobotType {
  HUMANOID = "humanoid",
  MOBILE_ROBOT = "mobile_robot",
  MANIPULATOR = "manipulator",
  AERIAL_DRONE = "aerial_drone",
  UNDERWATER_ROBOT = "underwater_robot",
  WHEELED_ROBOT = "wheeled_robot",
  TRACKED_ROBOT = "tracked_robot",
  CUSTOM = "custom"
}

export enum RobotStatusEnum {
  ONLINE = "online",
  OFFLINE = "offline",
  ERROR = "error",
  BUSY = "busy",
  IDLE = "idle",
  MAINTENANCE = "maintenance",
  SIMULATION = "simulation"
}

export enum ControlMode {
  POSITION = "position",
  VELOCITY = "velocity",
  TORQUE = "torque",
  IMPEDANCE = "impedance",
  FORCE = "force",
  TRAJECTORY = "trajectory"
}

export interface Robot {
  id: number;
  name: string;
  description?: string;
  robot_type: RobotType;
  model?: string;
  manufacturer?: string;
  status: RobotStatusEnum;
  last_seen?: string;
  battery_level: number;
  cpu_temperature?: number;
  connection_type: string;
  connection_params: Record<string, any>;
  ip_address?: string;
  port?: number;
  configuration: Record<string, any>;
  urdf_path?: string;
  simulation_engine: string;
  control_mode: ControlMode;
  capabilities: Record<string, any>;
  joint_count: number;
  sensor_count: number;
  user_id: number;
  is_public: boolean;
  is_default: boolean;
  created_at: string;
  updated_at: string;
}

export interface RobotJoint {
  id: number;
  robot_id: number;
  name: string;
  joint_type: string;
  min_position: number;
  max_position: number;
  max_velocity: number;
  max_torque: number;
  offset: number;
  direction: number;
  current_position: number;
  current_velocity: number;
  current_torque: number;
  temperature: number;
  description?: string;
  parent_link?: string;
  child_link?: string;
  axis: string;
  created_at: string;
  updated_at: string;
}

export interface RobotSensor {
  id: number;
  robot_id: number;
  name: string;
  sensor_type: string;
  model?: string;
  manufacturer?: string;
  sampling_rate?: number;
  accuracy?: number;
  range_min?: number;
  range_max?: number;
  position_x: number;
  position_y: number;
  position_z: number;
  orientation_x: number;
  orientation_y: number;
  orientation_z: number;
  orientation_w: number;
  status: string;
  last_data?: Record<string, any>;
  last_update?: string;
  description?: string;
  calibration_data?: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface RobotListResponse {
  robots: Robot[];
  total: number;
  skip: number;
  limit: number;
}

export interface RobotDetailResponse {
  robot: Robot;
  joints: RobotJoint[];
  sensors: RobotSensor[];
}

export interface ROSStatus {
  connected: boolean;
  uri: string;
  port: number;
  topics: string[];
  services: string[];
  last_connection: string | null;
  connection_duration: number;
}

export interface JointState {
  name: string;
  position: number;
  velocity: number;
  torque: number;
  temperature: number;
  voltage: number;
  current: number;
  timestamp: string;
}

export interface SensorData {
  type: string;
  name: string;
  unit: string;
  data: any;
  timestamp: string;
  accuracy: number;
  calibrated: boolean;
}

export interface RobotPose {
  pose_type: string;
  joint_positions: Record<string, number>;
  message: string;
  simulated: boolean;
  execution_time: number;
}

export interface RobotCapabilities {
  movement_control: boolean;
  joint_control: boolean;
  sensor_integration: boolean;
  ros_integration: boolean;
  gazebo_simulation: boolean;
  pybullet_simulation: boolean;
  motion_planning: boolean;
  collision_detection: boolean;
  hardware_abstraction: boolean;
  real_time_control: boolean;
  trajectory_generation: boolean;
  force_control: boolean;
  impedance_control: boolean;
  vision_based_control: boolean;
  multi_robot_coordination: boolean;
  autonomous_navigation: boolean;
  manipulation: boolean;
  locomotion: boolean;
}

export interface RobotApiResponse<T = any> {
  success: boolean;
  timestamp: string;
  data: T;
}

export const robotApi = {
  // 获取机器人状态
  async getStatus(): Promise<RobotApiResponse<RobotStatus>> {
    return apiClient.get('/robot/status');
  },

  // 获取特定机器人状态
  async getRobotStatus(robotId: number): Promise<RobotApiResponse<RobotStatus>> {
    return apiClient.get(`/robots/${robotId}/status`);
  },

  // ROS连接管理
  async connectROS(uri: string = 'http://localhost:11311', port: number = 9090): Promise<RobotApiResponse<{
    connected: boolean;
    uri: string;
    port: number;
    message: string;
  }>> {
    return apiClient.post('/robot/ros/connect', { uri, port });
  },

  async disconnectROS(): Promise<RobotApiResponse<{
    connected: boolean;
    message: string;
  }>> {
    return apiClient.post('/robot/ros/disconnect');
  },

  async getROSStatus(): Promise<RobotApiResponse<ROSStatus>> {
    return apiClient.get('/robot/ros/status');
  },

  // 关节控制
  async getJointStates(jointNames?: string[]): Promise<RobotApiResponse<{
    joints: JointState[];
    count: number;
    unit: Record<string, string>;
  }>> {
    const params = jointNames ? { joint_names: jointNames.join(',') } : {};
    return apiClient.get('/robot/joints', { params });
  },

  async controlJoint(
    jointName: string,
    controlMode: 'position' | 'velocity' | 'torque',
    position?: number,
    velocity?: number,
    torque?: number
  ): Promise<RobotApiResponse<{
    joint_name: string;
    control_mode: string;
    position?: number;
    velocity?: number;
    torque?: number;
    message: string;
    simulated: boolean;
  }>> {
    const params: any = { joint_name: jointName, control_mode: controlMode };
    if (position !== undefined) params.position = position;
    if (velocity !== undefined) params.velocity = velocity;
    if (torque !== undefined) params.torque = torque;
    
    return apiClient.post('/robot/joints/control', {}, { params });
  },

  // 传感器数据
  async getSensorData(sensorTypes?: string[]): Promise<RobotApiResponse<{
    sensors: SensorData[];
    count: number;
  }>> {
    const params = sensorTypes ? { sensor_types: sensorTypes.join(',') } : {};
    return apiClient.get('/robot/sensors', { params });
  },

  // 机器人姿态控制
  async setRobotPose(
    poseType: string,
    jointPositions?: Record<string, number>
  ): Promise<RobotApiResponse<RobotPose>> {
    const data: any = { pose_type: poseType };
    if (jointPositions) data.joint_positions = jointPositions;
    return apiClient.post('/robot/motion/pose', data);
  },

  // Gazebo控制
  async controlGazebo(
    action: 'start' | 'stop' | 'pause' | 'reset' | 'load_world',
    worldName?: string
  ): Promise<RobotApiResponse<{
    action: string;
    world_name?: string;
    message: string;
    simulated: boolean;
    gazebo_running: boolean;
  }>> {
    const params: any = { action };
    if (worldName) params.world_name = worldName;
    return apiClient.post('/robot/gazebo/control', {}, { params });
  },

  // PyBullet控制
  async controlPyBullet(
    action: 'connect' | 'disconnect' | 'step_simulation' | 'reset_simulation' | 'load_urdf',
    urdfPath?: string,
    physicsEngine?: string
  ): Promise<RobotApiResponse<{
    action: string;
    urdf_path?: string;
    physics_engine?: string;
    message: string;
    simulated: boolean;
    pybullet_connected: boolean;
  }>> {
    const params: any = { action };
    if (urdfPath) params.urdf_path = urdfPath;
    if (physicsEngine) params.physics_engine = physicsEngine;
    return apiClient.post('/robot/pybullet/control', {}, { params });
  },

  // 获取机器人能力
  async getCapabilities(): Promise<RobotApiResponse<{
    capabilities: RobotCapabilities;
    supported_count: number;
    total_count: number;
    compatibility: {
      ros_versions: string[];
      gazebo_versions: string[];
      simulation_engines: string[];
      robot_models: string[];
    };
  }>> {
    return apiClient.get('/robot/capabilities');
  },

  // 多机器人管理API
  async getRobots(
    skip: number = 0,
    limit: number = 100,
    robotType?: RobotType,
    statusFilter?: RobotStatusEnum
  ): Promise<RobotApiResponse<RobotListResponse>> {
    const params: any = { skip, limit };
    if (robotType) params.robot_type = robotType;
    if (statusFilter) params.status_filter = statusFilter;
    return apiClient.get('/robots/', { params });
  },

  async getRobot(robotId: number): Promise<RobotApiResponse<RobotDetailResponse>> {
    return apiClient.get(`/robots/${robotId}`);
  },

  async createRobot(
    name: string,
    robotType: RobotType = RobotType.HUMANOID,
    description?: string,
    model?: string,
    manufacturer?: string,
    configuration: Record<string, any> = {},
    urdfPath?: string,
    simulationEngine: string = 'gazebo',
    controlMode: ControlMode = ControlMode.POSITION
  ): Promise<RobotApiResponse<Robot>> {
    const params: any = {
      name,
      robot_type: robotType,
      description,
      model,
      manufacturer,
      configuration: JSON.stringify(configuration),
      urdf_path: urdfPath,
      simulation_engine: simulationEngine,
      control_mode: controlMode
    };
    // 移除undefined参数
    Object.keys(params).forEach(key => params[key] === undefined && delete params[key]);
    if (params.configuration) params.configuration = JSON.parse(params.configuration); // 重新解析为对象
    return apiClient.post('/robots/', {}, { params });
  },

  async connectRobot(robotId: number): Promise<RobotApiResponse<{
    robot_id: number;
    name: string;
    status: RobotStatusEnum;
    last_seen?: string;
  }>> {
    return apiClient.post(`/robots/${robotId}/connect`);
  },

  async disconnectRobot(robotId: number): Promise<RobotApiResponse<{
    robot_id: number;
    name: string;
    status: RobotStatusEnum;
  }>> {
    return apiClient.post(`/robots/${robotId}/disconnect`);
  },

  async setDefaultRobot(robotId: number): Promise<RobotApiResponse<Robot>> {
    return apiClient.post(`/robots/${robotId}/set-default`);
  },

  async getDefaultRobot(): Promise<RobotApiResponse<Robot>> {
    return apiClient.get('/robots/default');
  },

  // WebSocket连接（返回WebSocket URL）
  getWebSocketUrl(): string {
    const baseUrl = apiClient.getBaseUrl().replace('http://', 'ws://').replace('https://', 'wss://');
    return `${baseUrl}/robot/ws`;
  },

  // 运动控制API
  async planPath(
    startPosition: number[],
    goalPosition: number[],
    algorithm: string = 'astar',
    gridSize: number = 0.1,
    maxIterations: number = 1000,
    simulationType: string = 'pybullet'
  ): Promise<RobotApiResponse<{
    path: number[][];
    path_length: number;
    computation_time: number;
    nodes_explored: number;
    algorithm: string;
    simulation_type: string;
  }>> {
    return apiClient.post('/motion-control/plan-path', {
      start_position: startPosition,
      goal_position: goalPosition,
      algorithm,
      grid_size: gridSize,
      max_iterations: maxIterations,
      simulation_type: simulationType
    });
  },

  async executePath(
    path: number[][],
    speed: number = 0.1,
    simulationType: string = 'pybullet'
  ): Promise<RobotApiResponse<{
    points_executed: number;
    speed: number;
    simulation_type: string;
  }>> {
    return apiClient.post('/motion-control/execute-path', {
      path,
      speed,
      simulation_type: simulationType
    });
  },

  async moveToPosition(
    targetPosition: number[],
    speed: number = 0.1,
    simulationType: string = 'pybullet'
  ): Promise<RobotApiResponse<{
    target_position: number[];
    speed: number;
    simulation_type: string;
  }>> {
    return apiClient.post('/motion-control/move-to-position', {
      target_position: targetPosition,
      speed,
      simulation_type: simulationType
    });
  },

  async getSimulationInfo(
    simulationType: string = 'pybullet'
  ): Promise<RobotApiResponse<{
    initialized: boolean;
    connected: boolean;
    simulation_type: string;
    info: any;
  }>> {
    return apiClient.get('/motion-control/simulation-info', {
      params: { simulation_type: simulationType }
    });
  },

  // 示范学习API
  async startDemonstrationRecording(
    robotId: number,
    name: string,
    description: string = '',
    demonstrationType: string = 'joint_control',
    config: Record<string, any> = {}
  ): Promise<RobotApiResponse<{
    demonstration_id: number;
    name: string;
    message: string;
    started_at: string;
  }>> {
    return apiClient.post(`/robot/demonstration/${robotId}/start-recording`, {
      name,
      description,
      demonstration_type: demonstrationType,
      config
    });
  },

  async stopDemonstrationRecording(
    robotId: number
  ): Promise<RobotApiResponse<{
    demonstration_id: number;
    name: string;
    message: string;
    stopped_at: string;
    total_frames: number;
  }>> {
    return apiClient.post(`/robot/demonstration/${robotId}/stop-recording`);
  },

  async pauseDemonstrationRecording(
    robotId: number
  ): Promise<RobotApiResponse<{
    demonstration_id: number;
    name: string;
    message: string;
    paused_at: string;
    frames_recorded: number;
    recording_paused: boolean;
  }>> {
    return apiClient.post(`/robot/demonstration/${robotId}/pause-recording`);
  },

  async resumeDemonstrationRecording(
    robotId: number
  ): Promise<RobotApiResponse<{
    demonstration_id: number;
    name: string;
    message: string;
    resumed_at: string;
    frames_recorded: number;
    recording_resumed: boolean;
  }>> {
    return apiClient.post(`/robot/demonstration/${robotId}/resume-recording`);
  },

  async recordRobotState(
    robotId: number,
    jointPositions: Record<string, number> = {},
    jointVelocities?: Record<string, number>,
    jointTorques?: Record<string, number>,
    sensorData?: Record<string, any>,
    imuData?: Record<string, any>,
    controlCommands?: Record<string, any>,
    environmentState?: Record<string, any>
  ): Promise<RobotApiResponse<{
    frame_id: number;
    timestamp: string;
    message: string;
  }>> {
    return apiClient.post(`/robot/demonstration/${robotId}/record-state`, {
      joint_positions: jointPositions,
      joint_velocities: jointVelocities,
      joint_torques: jointTorques,
      sensor_data: sensorData,
      imu_data: imuData,
      control_commands: controlCommands,
      environment_state: environmentState
    });
  },

  async startDemonstrationPlayback(
    robotId: number,
    demonstrationId: number
  ): Promise<RobotApiResponse<{
    demonstration_id: number;
    robot_id: number;
    message: string;
    started_at: string;
  }>> {
    return apiClient.post(`/robot/demonstration/${robotId}/start-playback/${demonstrationId}`);
  },

  async pauseDemonstrationPlayback(
    robotId: number
  ): Promise<RobotApiResponse<{
    demonstration_id: number;
    robot_id: number;
    message: string;
    paused_at: string;
    playback_paused: boolean;
  }>> {
    return apiClient.post(`/robot/demonstration/${robotId}/pause-playback`);
  },

  async resumeDemonstrationPlayback(
    robotId: number
  ): Promise<RobotApiResponse<{
    demonstration_id: number;
    robot_id: number;
    message: string;
    resumed_at: string;
    playback_resumed: boolean;
  }>> {
    return apiClient.post(`/robot/demonstration/${robotId}/resume-playback`);
  },

  async stopDemonstrationPlayback(
    robotId: number
  ): Promise<RobotApiResponse<{
    demonstration_id: number;
    robot_id: number;
    message: string;
    stopped_at: string;
  }>> {
    return apiClient.post(`/robot/demonstration/${robotId}/stop-playback`);
  }
};

File: backend/db_models/robot.py
"""
机器人相关数据库模型
包含机器人配置、状态、硬件信息等模型
"""

from sqlalchemy import (
    Column,
    Integer,
    String,
    Boolean,
    DateTime,
    Float,
    Text,
    ForeignKey,
    JSON,
    Enum as SQLEnum,
)
from sqlalchemy.orm import relationship
from datetime import datetime
from enum import Enum as PyEnum

from ..core.database import Base


class RobotType(PyEnum):
    """机器人类型枚举"""

    HUMANOID = "humanoid"  # 人形机器人
    MOBILE_ROBOT = "mobile_robot"  # 移动机器人
    MANIPULATOR = "manipulator"  # 机械臂
    AERIAL_DRONE = "aerial_drone"  # 无人机
    UNDERWATER_ROBOT = "underwater_robot"  # 水下机器人
    WHEELED_ROBOT = "wheeled_robot"  # 轮式机器人
    TRACKED_ROBOT = "tracked_robot"  # 履带式机器人
    CUSTOM = "custom"  # 自定义类型


class RobotStatus(PyEnum):
    """机器人状态枚举"""

    ONLINE = "online"  # 在线
    OFFLINE = "offline"  # 离线
    ERROR = "error"  # 错误
    BUSY = "busy"  # 繁忙
    IDLE = "idle"  # 空闲
    MAINTENANCE = "maintenance"  # 维护中
    SIMULATION = "simulation"  # 仿真模式


class ControlMode(PyEnum):
    """控制模式枚举"""

    POSITION = "position"  # 位置控制
    VELOCITY = "velocity"  # 速度控制
    TORQUE = "torque"  # 力矩控制
    IMPEDANCE = "impedance"  # 阻抗控制
    FORCE = "force"  # 力控制
    TRAJECTORY = "trajectory"  # 轨迹跟踪


class Robot(Base):
    """机器人主表"""

    __tablename__ = "robots"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(255), nullable=False, unique=True, comment="机器人名称")
    description = Column(Text, comment="机器人描述")

    # 类型信息
    robot_type = Column(
        SQLEnum(RobotType),
        nullable=False,
        default=RobotType.HUMANOID,
        comment="机器人类型",
    )
    model = Column(String(255), comment="机器人型号")
    manufacturer = Column(String(255), comment="制造商")

    # 状态信息
    status = Column(
        SQLEnum(RobotStatus),
        nullable=False,
        default=RobotStatus.OFFLINE,
        comment="机器人状态",
    )
    last_seen = Column(DateTime, comment="最后在线时间")
    battery_level = Column(Float, default=100.0, comment="电池电量 (%)")
    cpu_temperature = Column(Float, comment="CPU温度 (°C)")

    # 连接信息
    connection_type = Column(
        String(50),
        default="simulation",
        comment="连接类型: simulation, serial, ethernet, wifi, bluetooth",
    )
    connection_params = Column(JSON, default=dict, comment="连接参数")
    ip_address = Column(String(45), comment="IP地址")
    port = Column(Integer, comment="端口号")

    # 配置信息
    configuration = Column(JSON, default=dict, comment="机器人配置")
    urdf_path = Column(String(500), comment="URDF模型路径")
    simulation_engine = Column(
        String(50), default="gazebo", comment="仿真引擎: gazebo, pybullet, vrep"
    )
    control_mode = Column(
        SQLEnum(ControlMode), default=ControlMode.POSITION, comment="控制模式"
    )

    # 硬件能力
    capabilities = Column(JSON, default=dict, comment="机器人能力配置")
    joint_count = Column(Integer, default=0, comment="关节数量")
    sensor_count = Column(Integer, default=0, comment="传感器数量")

    # 权限和归属
    user_id = Column(
        Integer, ForeignKey("users.id"), nullable=False, comment="所属用户ID"
    )
    is_public = Column(Boolean, default=False, comment="是否公开")
    is_default = Column(Boolean, default=False, comment="是否默认机器人")

    # 时间戳
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False
    )

    # 关系
    user = relationship("User", back_populates="robots")

    def to_dict(self):
        """转换为字典"""
        return {
            "id": self.id,
            "name": self.name,
            "description": self.description,
            "robot_type": self.robot_type.value,
            "model": self.model,
            "manufacturer": self.manufacturer,
            "status": self.status.value,
            "last_seen": self.last_seen.isoformat() if self.last_seen else None,
            "battery_level": self.battery_level,
            "cpu_temperature": self.cpu_temperature,
            "connection_type": self.connection_type,
            "connection_params": self.connection_params,
            "ip_address": self.ip_address,
            "port": self.port,
            "configuration": self.configuration,
            "urdf_path": self.urdf_path,
            "simulation_engine": self.simulation_engine,
            "control_mode": self.control_mode.value,
            "capabilities": self.capabilities,
            "joint_count": self.joint_count,
            "sensor_count": self.sensor_count,
            "user_id": self.user_id,
            "is_public": self.is_public,
            "is_default": self.is_default,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
        }


class RobotJoint(Base):
    """机器人关节表"""

    __tablename__ = "robot_joints"

    id = Column(Integer, primary_key=True, index=True)
    robot_id = Column(
        Integer, ForeignKey("robots.id"), nullable=False, comment="所属机器人ID"
    )
    name = Column(String(100), nullable=False, comment="关节名称")
    joint_type = Column(
        String(50), default="revolute", comment="关节类型: revolute, prismatic, fixed"
    )

    # 物理参数
    min_position = Column(Float, default=-3.14159, comment="最小位置 (rad)")
    max_position = Column(Float, default=3.14159, comment="最大位置 (rad)")
    max_velocity = Column(Float, default=1.0, comment="最大速度 (rad/s)")
    max_torque = Column(Float, default=10.0, comment="最大扭矩 (N·m)")

    # 校准参数
    offset = Column(Float, default=0.0, comment="零点偏移 (rad)")
    direction = Column(Float, default=1.0, comment="方向 (+1或-1)")

    # 状态信息
    current_position = Column(Float, default=0.0, comment="当前位置 (rad)")
    current_velocity = Column(Float, default=0.0, comment="当前速度 (rad/s)")
    current_torque = Column(Float, default=0.0, comment="当前扭矩 (N·m)")
    temperature = Column(Float, default=25.0, comment="温度 (°C)")

    # 元数据
    description = Column(Text, comment="关节描述")
    parent_link = Column(String(100), comment="父连杆")
    child_link = Column(String(100), comment="子连杆")
    axis = Column(String(50), default="z", comment="关节轴: x, y, z")

    # 时间戳
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False
    )

    # 关系
    robot = relationship("Robot", back_populates="joints")

    def to_dict(self):
        """转换为字典"""
        return {
            "id": self.id,
            "robot_id": self.robot_id,
            "name": self.name,
            "joint_type": self.joint_type,
            "min_position": self.min_position,
            "max_position": self.max_position,
            "max_velocity": self.max_velocity,
            "max_torque": self.max_torque,
            "offset": self.offset,
            "direction": self.direction,
            "current_position": self.current_position,
            "current_velocity": self.current_velocity,
            "current_torque": self.current_torque,
            "temperature": self.temperature,
            "description": self.description,
            "parent_link": self.parent_link,
            "child_link": self.child_link,
            "axis": self.axis,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
        }


class RobotSensor(Base):
    """机器人传感器表"""

    __tablename__ = "robot_sensors"

    id = Column(Integer, primary_key=True, index=True)
    robot_id = Column(
        Integer, ForeignKey("robots.id"), nullable=False, comment="所属机器人ID"
    )
    name = Column(String(100), nullable=False, comment="传感器名称")
    sensor_type = Column(
        String(50),
        nullable=False,
        comment="传感器类型: imu, camera, lidar, force, touch",
    )

    # 配置信息
    model = Column(String(100), comment="传感器型号")
    manufacturer = Column(String(100), comment="制造商")
    sampling_rate = Column(Float, comment="采样率 (Hz)")
    accuracy = Column(Float, comment="精度")
    range_min = Column(Float, comment="最小量程")
    range_max = Column(Float, comment="最大量程")

    # 位置和方向
    position_x = Column(Float, default=0.0, comment="X位置 (m)")
    position_y = Column(Float, default=0.0, comment="Y位置 (m)")
    position_z = Column(Float, default=0.0, comment="Z位置 (m)")
    orientation_x = Column(Float, default=0.0, comment="X方向 (quaternion)")
    orientation_y = Column(Float, default=0.0, comment="Y方向 (quaternion)")
    orientation_z = Column(Float, default=0.0, comment="Z方向 (quaternion)")
    orientation_w = Column(Float, default=1.0, comment="W方向 (quaternion)")

    # 状态信息
    status = Column(
        String(50), default="online", comment="状态: online, offline, error"
    )
    last_data = Column(JSON, comment="最后数据")
    last_update = Column(DateTime, comment="最后更新时间")

    # 元数据
    description = Column(Text, comment="传感器描述")
    calibration_data = Column(JSON, comment="校准数据")

    # 时间戳
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(
        DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False
    )

    # 关系
    robot = relationship("Robot", back_populates="sensors")

    def to_dict(self):
        """转换为字典"""
        return {
            "id": self.id,
            "robot_id": self.robot_id,
            "name": self.name,
            "sensor_type": self.sensor_type,
            "model": self.model,
            "manufacturer": self.manufacturer,
            "sampling_rate": self.sampling_rate,
            "accuracy": self.accuracy,
            "range_min": self.range_min,
            "range_max": self.range_max,
            "position_x": self.position_x,
            "position_y": self.position_y,
            "position_z": self.position_z,
            "orientation_x": self.orientation_x,
            "orientation_y": self.orientation_y,
            "orientation_z": self.orientation_z,
            "orientation_w": self.orientation_w,
            "status": self.status,
            "last_data": self.last_data,
            "last_update": self.last_update.isoformat() if self.last_update else None,
            "description": self.description,
            "calibration_data": self.calibration_data,
            "created_at": self.created_at.isoformat(),
            "updated_at": self.updated_at.isoformat(),
        }


# 在Robot模型中添加关系
Robot.joints = relationship(
    "RobotJoint", back_populates="robot", cascade="all, delete-orphan"
)
Robot.sensors = relationship(
    "RobotSensor", back_populates="robot", cascade="all, delete-orphan"
)
Robot.demonstrations = relationship(
    "Demonstration", back_populates="robot", cascade="all, delete-orphan"
)
Robot.demonstration_tasks = relationship(
    "DemonstrationTask", back_populates="robot", cascade="all, delete-orphan"
)
Robot.market_listings = relationship(
    "RobotMarketListing", back_populates="robot", cascade="all, delete-orphan"
)

# 在User模型中需要添加robots关系（需要更新user.py）
# 我们将在初始化数据库时处理
